client/state/sites/reducer.js

```diff
@@ -138,34 +138,49 @@ export function items( state = {}, action ) {
 
 			// A site settings update may or may not include the icon property.
 			// If not, we should simply return state unchanged.
-			if ( ! settings.hasOwnProperty( 'site_icon' ) ) {
+			if ( ! settings.hasOwnProperty( 'site_icon' ) && ! settings.hasOwnProperty( 'blog_public' ) ) {
```

Comment:
Not sure what gotchas you might encounter, but doing a `reduce` on the keys we're monitoring might be a nicer way to encapsulate the separate logic.

Something like:

```js
return reduce( [ 'blog_public', 'site_icon' ], ( memo, key ) => {
	if ( ! settings.hasOwnProperty( key ) ) {
		return memo;
	}

	switch ( key ) {
		case 'blog_public':
			// ...

		case 'site_icon':
			// ...
	}

	return memo;
}, state );
```

(`SITE_RECEIVE` reducer above is similar in how it reduces using state as initial value while avoiding direct mutations)